portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,27 +14,57 @@
 
 package com.google.sps.servlets;
 
+import com.google.gson.Gson;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 import java.io.IOException;
-import javax.servlet.ServletException;
 import java.io.UnsupportedEncodingException;
-import com.google.gson.Gson;
+import java.sql.Timestamp;
 import java.util.*; 
+import javax.servlet.ServletException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import java.nio.charset.StandardCharsets;
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+/** Servlet that handles comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
-
-  private List<String> comments = new ArrayList<String>();
   
+  public static final int MAX_COMMENTS = 100;
+  private DatastoreService datastore;
+  private Query commentsQuery;
+  /** Initializes data needed to load comments from datastore when requested */
+  @Override
+  public void init() {
+    datastore = DatastoreServiceFactory.getDatastoreService();
+    commentsQuery = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+  }
+
+  /**
+   * Sends at most commentsLimit comments as a JSON string
+   */
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Integer limit = null;
+    try {
+      // Get the maximum number of comments to be displayed from the queryString
+      limit = Integer.parseInt(request.getParameter("commentsLimit"));
+      if (limit < 0 || limit > MAX_COMMENTS) {
```

Comment:
So requesting zero is ok?  Maybe `limit <=0` if it is not?

Reply:
To me, it seems ok to show 0 comments. Would it be better to use limit <=0 and have a separate button such as "Hide comments"?

Reply:
I think it is important to keep in mind which part of the system is doing what here - what is the java responsible for, and what is the javascript responsible for?

So should it be the responsibility for the backend server (i.e. the Java code) to hide all comments, or the client code (i.e. the Javascript code) to hide the comments?

Up to you, but for me it makes sense for the backend *only* to be concerned with returning data, and the javascript can just show or hide things depending on what the user is doing.  By blurring the two parts here ("make a server request for zero comments to hide something in the user's browser") it makes things harder to reuse/maintain later because now the backend server is not just returning data, but its behavior is intertwined with what happens in the browser too.

... and after all it just feels kinda weird making a request to the backend for nothing! :)

Reply:
You're right, I will change it to limit <= 0 and consider this approach for the next PRs.